# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [1]:
# Load python packages
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime
import seaborn as sns

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [2]:
os.chdir("/Users/fatimasoytemiz/Desktop/Springboard/Course/Unit 6/Data")
df = pd.read_csv("step3_output.csv")
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...              200.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              200.0   
4        0.0          0          1     2  ...              200.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [3]:
# Create a set of dummy variables from the state variable
df1 = pd.concat([df, pd.get_dummies(df['state'])], axis=1)
df1 = df1.drop(["state"], axis=1)
df1.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [4]:
# Scale numeric features
from sklearn import preprocessing
X = df1.drop(['Name','AdultWeekend'], axis=1)
y = df1.AdultWeekend
scale = preprocessing.StandardScaler().fit(X)
X_scale=scale.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [5]:
from sklearn.model_selection import train_test_split
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scale, y, test_size=0.25, random_state=1)

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [7]:
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed.

In [8]:
explained_variance_score(y_test, y_pred)

0.9311393710203248

In [9]:
mean_absolute_error(y_test, y_pred)

5.13161152189798

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [10]:
lm.intercept_

64.14857129643494

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [11]:
coeff_df1 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df1.sort_values(by=['Coefficient'],ascending=False).head(10).abs()

Coefficient
total_chairs  1.936290e+13
New York      7.384313e+12
fastQuads     7.340519e+12
Michigan      6.968786e+12
surface       6.877516e+12
Colorado      6.139904e+12
double        6.060722e+12
California    6.008468e+12
Pennsylvania  5.733659e+12
triple        5.406580e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [12]:
X = df.drop(['AdultWeekend','Name','state'], axis=1)
y = df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scale = scaler.transform(X) 
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scale, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.9215867687120808
5.5659047581659395


In [13]:
coeff_df2 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df2.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday        20.220173
summit_elev         15.284646
base_elev           11.565961
vertical_drop        4.764909
averageSnowfall      2.105686
quad                 1.486400
triple               1.395455
daysOpenLastYear     1.188288
surface              1.115737
Runs                 1.078848

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [14]:
X = df.drop(['Name','AdultWeekend','state','summit_elev','base_elev'], axis=1)
y = df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scale = scaler.transform(X) 
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scale, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.9280203667942301
5.432209860793244


In [15]:
coeff_df3 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df3.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday        20.132276
averageSnowfall      2.059629
quad                 1.494516
triple               1.347506
Runs                 1.273515
vertical_drop        1.197117
surface              1.104625
daysOpenLastYear     1.086438
fastQuads            0.843248
clusters             0.798635

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9311 | 5.1316 |-|
| Model 2. | 0.9216|5.5659 |'state'|
| Model 3. | 0.9280 |5.4322 |'state','summit_elev','base_elev'|

Model Selection:

In [16]:
# Model 3 has a better performance by not including correlated features.
